### PR TITLE
Issue 267

### PR DESCRIPTION
**Pull request by bennahugo**
- Exposes field, antenna and scan subset selections as switches
- Switch to argparse because optparse is deprecated and does not have support for custom types

The following is an example:
```
(venvkatdal) > $ mvftoms.py --antenna m000,m001 --scans 0~5 --field J1939-6342,J1331+3030 --flags cam,data_lost -f 1559937657.rdb
No calibration products will be applied
Per user request the following antennae will be dumped: 'm000', 'm001'
Per user request the following fields will be dumped: 'J1939-6342', 'J1331+3030'
Per user request the following scans will be dumped: 1, 3
Extract MS for spw 0: centre frequency 1284000000 Hz
Will create MS output in 1559937657.ms

#### Producing a full polarisation MS (HH,HV,VH,VV) ####


Using array as the reference antenna. All targets and scans will be based on this antenna.


Iterating through scans in dataset(s)...

Writing static meta data...
scan   1 ( 151 samples) loaded. Target: 'J1939-6342'. Writing to disk...
Wrote scan data (99.154664 MiB) in 111.232912 s (0.891415 MiBps)

scan   3 ( 601 samples) loaded. Target: 'J1331+3030'. Writing to disk...
Wrote scan data (394.648696 MiB) in 442.600555 s (0.891659 MiBps)


Writing dynamic fields to disk....

```  

and a listobs as follows:
```
2019-10-02 20:11:34     INFO    listobs::::+    ##########################################
2019-10-02 20:11:34     INFO    listobs::::+    ##### Begin Task: listobs            #####
2019-10-02 20:11:34     INFO    listobs::::     listobs(vis="1559937657.ms/",selectdata=True,spw="",field="",antenna="",
2019-10-02 20:11:34     INFO    listobs::::+            uvrange="",timerange="",correlation="",scan="",intent="",
2019-10-02 20:11:34     INFO    listobs::::+            feed="",array="",observation="",verbose=True,listfile="",
2019-10-02 20:11:34     INFO    listobs::::+            listunfl=False,cachesize=50,overwrite=False)
2019-10-02 20:11:34     INFO    listobs::ms::summary    ================================================================================
2019-10-02 20:11:34     INFO    listobs::ms::summary+              MeasurementSet Name:  /scratch/bhugo/CLUSTER_SURVEY_TAKE2/1559937657.ms      MS Version 2
2019-10-02 20:11:34     INFO    listobs::ms::summary+   ================================================================================
2019-10-02 20:11:34     INFO    listobs::ms::summary+      Observer: Marisa Geyer     Project: 20190607-0020  
2019-10-02 20:11:34     INFO    listobs::ms::summary+   Observation: MeerKAT
2019-10-02 20:11:34     INFO    listobs::MSMetaData::_computeScanAndSubScanProperties   Computing scan and subscan properties...
2019-10-02 20:11:35     INFO    listobs::ms::summary    Data records: 2256       Total elapsed time = 1595.33 seconds
2019-10-02 20:11:35     INFO    listobs::ms::summary+      Observed from   07-Jun-2019/20:02:26.4   to   07-Jun-2019/20:29:01.7 (UTC)
2019-10-02 20:11:35     INFO    listobs::ms::summary
2019-10-02 20:11:35     INFO    listobs::ms::summary+      ObservationID = 0         ArrayID = 0
2019-10-02 20:11:35     INFO    listobs::ms::summary+     Date        Timerange (UTC)          Scan  FldId FieldName             nRows     SpwIds   Average Interval(s)    ScanIntent
2019-10-02 20:11:35     INFO    listobs::ms::summary+     07-Jun-2019/20:02:26.4 - 20:07:28.3     1      0 J1939-6342                 453  [0]  [2] [CALIBRATE_BANDPASS,CALIBRATE_FLUX]
2019-10-02 20:11:35     INFO    listobs::ms::summary+                 20:09:00.2 - 20:29:01.7     2      1 J1331+3030                1803  [0]  [2] [CALIBRATE_BANDPASS,CALIBRATE_FLUX]
2019-10-02 20:11:35     INFO    listobs::ms::summary               (nRows = Total number of rows per scan) 
2019-10-02 20:11:35     INFO    listobs::ms::summary    Fields: 2
2019-10-02 20:11:35     INFO    listobs::ms::summary+     ID   Code Name                RA               Decl           Epoch   SrcId      nRows
2019-10-02 20:11:35     INFO    listobs::ms::summary+     0    T    J1939-6342          19:39:25.050000 -63.42.43.60000 J2000   0            453
2019-10-02 20:11:35     INFO    listobs::ms::summary+     1    T    J1331+3030          13:31:08.290000 +30.30.33.00000 J2000   1           1803
2019-10-02 20:11:35     INFO    listobs::ms::summary    Spectral Windows:  (1 unique spectral windows and 1 unique polarization setups)
2019-10-02 20:11:35     INFO    listobs::ms::summary+     SpwID  Name   #Chans   Frame   Ch0(MHz)  ChanWid(kHz)  TotBW(kHz) CtrFreq(MHz)  Corrs          
2019-10-02 20:11:35     INFO    listobs::ms::summary+     0      none    4096   TOPO     856.000       208.984    856000.0   1283.8955   XX  XY  YX  YY
2019-10-02 20:11:35     INFO    listobs::ms::summary    Sources: 2
2019-10-02 20:11:35     INFO    listobs::ms::summary+     ID   Name                SpwId RestFreq(MHz)  
2019-10-02 20:11:35     INFO    listobs::ms::summary+     0    J1939-6342          0     -              
2019-10-02 20:11:35     INFO    listobs::ms::summary+     1    J1331+3030          0     -              
2019-10-02 20:11:35     INFO    listobs::ms::summary+     NB: No systemic velocity information found in SOURCE table.
2019-10-02 20:11:35     INFO    listobs::ms::summary    Antennas: 2:
2019-10-02 20:11:35     INFO    listobs::ms::summary+     ID   Name  Station   Diam.    Long.         Lat.                Offset from array center (m)                ITRF Geocentric coordinates (m)        
2019-10-02 20:11:35     INFO    listobs::ms::summary+                                                                        East         North     Elevation               x               y               z
2019-10-02 20:11:35     INFO    listobs::ms::summary+     0    m000  m000      13.5 m   +021.26.37.7  -30.32.39.3         -9.3284       30.3604       45.2895  5109271.497354  2006808.893028 -3239130.736141
2019-10-02 20:11:35     INFO    listobs::ms::summary+     1    m001  m001      13.5 m   +021.26.38.0  -30.32.38.1          0.0560       65.8884       45.2670  5109284.854078  2006824.221724 -3239100.126460
2019-10-02 20:11:35     INFO    listobs::::     ##### End Task: listobs              #####
2019-10-02 20:11:35     INFO    listobs::::+    ##########################################
```

### CODE REVIEW

**Comment by bennahugo** [comment]
This is from the following superset:
```
===============================================================================
Name: file:///1559937657.rdb | 1559937657-sdp-l0 (version 4.0)
===============================================================================
Observer: Marisa Geyer  Experiment ID: 20190607-0020
Description: 'A galaxy cluster safari with MeerKAT - 2s'
Observed from 2019-06-07 22:02:24.387 SAST to 2019-06-08 12:10:28.860 SAST
Dump rate / period: 0.50021 Hz / 1.999 s
Subarrays: 1
  ID  Antennas                            Inputs  Corrprods
   0  m000,m001,m002,m003,m004,m005,m006,m007,m008,m009,m010,m011,m012,m013,m014,m015,m016,m017,m018,m019,m020,m021,m022,m023,m024,m025,m026,m027,m028,m029,m030,m031,m032,m033,m034,m035,m037,m038,m040,m041,m042,m043,m044,m045,m046,m047,m048,m049,m050,m051,m052,m053,m0
54,m055,m056,m057,m058,m059,m060,m061,m062,m063  124      7812
Spectral Windows: 1
  ID Band Product  CentreFreq(MHz)  Bandwidth(MHz)  Channels  ChannelWidth(kHz)
   0 L    c856M4k    1284.000         856.000           4096       208.984
-------------------------------------------------------------------------------
Data selected according to the following criteria:
  ants=['m000', 'm001', 'm002', 'm003', 'm004', 'm005', 'm006', 'm007', 'm008', 'm009', 'm010', 'm011', 'm012', 'm013', 'm014', 'm015', 'm016', 'm017', 'm018', 'm019', 'm020', 'm021', 'm022', 'm023', 'm024', 'm025', 'm026', 'm027', 'm028', 'm029', 'm030', 'm031', 'm03
2', 'm033', 'm034', 'm035', 'm037', 'm038', 'm040', 'm041', 'm042', 'm043', 'm044', 'm045', 'm046', 'm047', 'm048', 'm049', 'm050', 'm051', 'm052', 'm053', 'm054', 'm055', 'm056', 'm057', 'm058', 'm059', 'm060', 'm061', 'm062', 'm063']
  spw=0 
  subarray=0
-------------------------------------------------------------------------------
Shape: (25453 dumps, 4096 channels, 7812 correlation products) => Size: 6515.551 GB
Antennas: m000,m001,m002,m003,m004,m005,m006,m007,m008,m009,m010,m011,m012,m013,m014,m015,m016,m017,m018,m019,m020,m021,m022,m023,m024,m025,m026,m027,m028,m029,m030,m031,m032,m033,m034,m035,m037,m038,m040,m041,m042,m043,m044,m045,m046,m047,m048,m049,m050,m051,m052,m05
3,m054,m055,m056,m057,m058,m059,m060,m061,m062,m063  Inputs: 124  Autocorr: yes  Crosscorr: yes
Channels: 4096 (index 0 - 4095,  856.000 MHz - 1711.791 MHz), each 208.984 kHz wide
Targets: 19 selected out of 19 in catalogue
  ID  Name               Type      RA(J2000)     DEC(J2000)  Tags           Dumps  ModelFlux(Jy)
   0  J1939-6342         radec     19:39:25.05  -63:42:43.6  bpcal fluxcal   1013
   1  J1331+3030         radec     13:31:08.29   30:30:33.0  bpcal            647
   2  J1120-2508         radec     11:20:09.14  -25:08:07.2  gaincal          262
   3  A 1300             radec     11:31:54.40  -19:55:42.0  target          1820
   4  J1733-1304         radec     17:33:02.66  -13:04:49.1  gaincal          317
   5  PSZ1G 018.75+23.5  radec     17:02:22.10   -1:00:16.0  target          1836
   6  J1859-6615         radec     18:59:57.80  -66:15:02.8  gaincal          294
   7  RXCJ 1638.2-6420   radec     16:38:18.30  -64:21:07.0  target          1837
   8  J1619-8418         radec     16:19:33.97  -84:18:19.1  gaincal          252
   9  RXCJ 1631.6-7507   radec     16:31:24.00  -75:07:01.0  target          1831
  10  J1951-2737         radec     19:51:23.26  -27:37:20.5  gaincal          296
  11  MACSJ 1931-2635    radec     19:31:49.60  -26:34:34.0  target          1829
  12  J0408-6545         radec      4:08:20.38  -65:45:09.6  bpcal fluxcal    701
  13  J0303-6211         radec      3:03:50.81  -62:11:25.4  gaincal          510
  14  PSZ2G277.76-51.74  radec      2:54:17.50  -58:57:10.8  target          3639
  15  J0203-4349         radec      2:03:40.75  -43:49:50.8  gaincal          532
  16  PSZ2G254.08-58.45  radec      3:04:16.80  -44:01:51.6  target          3669
  17  J0440-4333         radec      4:40:17.07  -43:33:09.0  gaincal          510
  18  PSZ2G255.60-46.18  radec      4:11:15.50  -48:19:19.2  target          3658
Scans: 183 selected out of 183 total       Compscans: 77 selected out of 77 total
  Date        Timerange(UTC)       ScanState  CompScanLabel  Dumps  Target
  07-Jun-2019/20:02:25 - 20:02:25    0:slew     0:track          1    0:J1939-6342
              20:02:27 - 20:07:27    1:track    0:track        151    0:J1939-6342
              20:07:29 - 20:08:59    2:slew     1:track         46    1:J1331+3030
              20:09:01 - 20:29:00    3:track    1:track        601    1:J1331+3030
              20:29:02 - 20:29:46    4:slew     2:track         23    2:J1120-2508
              20:29:48 - 20:32:18    5:track    2:track         76    2:J1120-2508
...
```

**Comment by bmerry** [comment]
I'm assigning this over to @ludwigschwardt as he's the katdal maintainer.

**Comment by bennahugo** [comment]
ping @ludwigschwardt can you please review

File: scripts/mvftoms.py [hunk 1]
```diff
@@ -143,76 +147,119 @@ def main():
                      'bpcal': 'CALIBRATE_BANDPASS,CALIBRATE_FLUX',
                      'target': 'TARGET'}
 
-    usage = "%prog [options] <dataset> [<dataset2>]*"
+    def casa_style_int_list(val, argparse=False, opt_unit="m"):
```
Comment (ludwigschwardt) [inline scripts/mvftoms.py]: What is the point of `opt_unit` for antenna lists?

File: scripts/mvftoms.py [hunk 2]
```diff
@@ -143,76 +147,119 @@ def main():
...
+        """ returns list of ints """
+        RangeException = ArgumentTypeError if argparse else ValueError
+        if val.strip() == "" or val.strip() == "*":
+            return None
+        elif re.match(r"^(\d+)(~\d+[" + opt_unit +
```
Comment (ludwigschwardt) [inline scripts/mvftoms.py]: It seems like a bit of a waste to do a full regexp and then parse the string using split, filter, replace, etc.

How about simplifying both approaches by splitting the string on commas first and then doing the following regexp?

```Python
int_or_range = r"^(\d+)(~(\d+)[{}]?)?$".format(opt_unit)
match = re.match(int_or_range, val)
if not match:
    raise RangeException(...)
if match.group(3):
    # range
    rmin = int(match.group(1))
    rmax = int(match.group(3))
    vals.append(list(range(rmin, rmax + 1))
else:
    # single int
    vals.append(int(match.group(1))
```

File: scripts/mvftoms.py [hunk 3]
```diff
@@ -281,6 +328,77 @@ def _cp_index(a1, a2, pol):
...
+                         "You requested {0:s}, but only {1:s} are available.".format(
+                         ", ".join(["'{}'".format(f) for f in dump_ants]),
+                         ", ".join(["'{}'".format(f) for f in avail_ants])))
+
+    if len(dump_ants) == 0:
+       print('User antenna criterion resulted in empty database, nothing to be done. '
+             'Perhaps you wanted to select from the following: {}'.format(
+             ", ".join(["'{}'".format(f) for f in avail_ants])))
+
+    print('Per user request the following antennae will be dumped: {}'.format(
```
Comment (ludwigschwardt) [inline scripts/mvftoms.py]: antennae, dumped...

File: scripts/mvftoms.py [hunk 4]
```diff
@@ -281,6 +328,77 @@ def _cp_index(a1, a2, pol):
...
+    # be specified multiple times. Put in a guard to split comma separated lists as well
+    split_ants = []
+    for a in dump_ants:
+        split_ants += [x.strip() for x in a.split(",")]
+    dump_ants = split_ants
+
+    if not set(dump_ants) <= set(avail_ants):
+        raise ValueError("One or more antennas cannot be found in the dataset. "
+                         "You requested {0:s}, but only {1:s} are available.".format(
+                         ", ".join(["'{}'".format(f) for f in dump_ants]),
```
Comment (ludwigschwardt) [inline scripts/mvftoms.py]: `repr(f)` is simpler than `"'{}'".format(f)`, although for antenna names the quotation marks will be a distraction.

File: scripts/mvftoms.py [hunk 5]
```diff
@@ -281,6 +328,77 @@ def _cp_index(a1, a2, pol):
...
+    # be specified multiple times. Put in a guard to split comma separated lists as well
+    split_fields = []
+    for f in dump_fields:
+        split_fields += [x.strip() for x in f.split(",")]
+    dump_fields = split_fields
+
+    if not set(dump_fields) <= set(avail_fields):
+        raise ValueError("One or more fields cannot be found in the dataset. "
+                         "You requested {0:s}, but only {1:s} are available.".format(
+                         ", ".join(["'{}'".format(f) for f in dump_fields]),
```
Comment (ludwigschwardt) [inline scripts/mvftoms.py]: `repr(f)` is simpler than `"'{}'".format(f)`.

File: scripts/mvftoms.py [hunk 6]
```diff
@@ -281,6 +328,77 @@ def _cp_index(a1, a2, pol):
...
+
+    print('Per user request the following fields will be dumped: {}'.format(
+        ", ".join(["'{}'".format(f) for f in dump_fields])))
+
+    dataset.select(targets=dump_fields)
+
+    # get a set of user selected available tracking scans, ignore slew scans
+    avail_tracks = map(lambda x: x[0],
+                     filter(lambda x: x[1] == 'track',
+                            list(dataset.scans())))
```
Comment (ludwigschwardt) [inline scripts/mvftoms.py]: This `list` seems superfluous.

File: scripts/mvftoms.py [hunk 7]
```diff
@@ -281,6 +328,77 @@ def _cp_index(a1, a2, pol):
...
+    dump_scans = list(set(dump_scans).intersection(set(avail_tracks)))
+    if len(dump_scans) == 0:
+        print('User scan criterion resulted in empty database, nothing to be done. '
+              'Perhaps you wanted to select from the following: {}'.format(
+               ", ".join(map(str, avail_tracks))))
+        sys.exit(0)
+    print('Per user request the following scans will be dumped: {}'.format(
+        ", ".join(map(str, dump_scans))))
+
+    dataset.select(scans=dump_scans, targets=dump_fields)
```
Comment (ludwigschwardt) [inline scripts/mvftoms.py]: This select seems superfluous.

File: scripts/mvftoms.py [hunk 8]
```diff
@@ -281,6 +328,77 @@ def _cp_index(a1, a2, pol):
...
+
+    # some users may specify comma-separated lists although we said the switch should
+    # be specified multiple times. Put in a guard to split comma separated lists as well
+    split_fields = []
+    for f in dump_fields:
+        split_fields += [x.strip() for x in f.split(",")]
+    dump_fields = split_fields
+
+    if not set(dump_fields) <= set(avail_fields):
+        raise ValueError("One or more fields cannot be found in the dataset. "
```
Comment (ludwigschwardt) [inline scripts/mvftoms.py]: I think we use `RuntimeError` for these kinds of errors in the rest of the script.

File: scripts/mvftoms.py [hunk 9]
```diff
@@ -281,6 +328,77 @@ def _cp_index(a1, a2, pol):
...
+                            list(dataset.scans())))
+
+    dump_scans = options.scans if options.scans else \
+                 avail_tracks
+    dump_scans = list(set(dump_scans).intersection(set(avail_tracks)))
+    if len(dump_scans) == 0:
+        print('User scan criterion resulted in empty database, nothing to be done. '
+              'Perhaps you wanted to select from the following: {}'.format(
+               ", ".join(map(str, avail_tracks))))
+        sys.exit(0)
```
Comment (ludwigschwardt) [inline scripts/mvftoms.py]: I would raise RuntimeError instead of a lone `sys.exit(0)`.

File: scripts/mvftoms.py [hunk 10]
```diff
@@ -143,76 +147,119 @@ def main():
...
+            raise RangeException("Value must be range, comma list or blank")
+
+    usage = "%(prog)s [options] <dataset> [<dataset2>]*"
     description = "Convert MVF dataset(s) to CASA MeasurementSet. The datasets may " \
                   "be local filenames or archive URLs (including access tokens). " \
                   "If there are multiple datasets they will be concatenated via " \
                   "katdal before conversion."
-    parser = optparse.OptionParser(usage=usage, description=description)
-    parser.add_option("-o", "--output-ms", default=None,
+    parser = argparse.ArgumentParser(usage=usage, description=description)
```
Comment (ludwigschwardt) [inline scripts/mvftoms.py]: Thanks for this!

**Comment by ludwigschwardt** [comment]
Thanks for argparse! Although optparse does have custom types (just implemented one 2 months ago!).

**Comment by bennahugo** [comment]
O.o couldn't quite get it to work in optparse - must have not tried hard enough :) Thanks I will comment above and change as needed

File: scripts/mvftoms.py (hunk 9 above)
Comment (bennahugo) [reply scripts/mvftoms.py]: Ok I thought this should not be cause for an exception. I will change it.

File: scripts/mvftoms.py (hunk 1 above)
Comment (bennahugo) [reply scripts/mvftoms.py]: This is a generic range parser. We could also possibly select only ranges of baselines to dump in the future?

Comment (bennahugo) [reply scripts/mvftoms.py]: opt unit can be a list of symbols to append, e.g. 0~100[m...]?

File: scripts/mvftoms.py (hunk 2 above)
Comment (bennahugo) [reply scripts/mvftoms.py]: neater... will change it

File: scripts/mvftoms.py (hunk 7 above)
Comment (bennahugo) [reply scripts/mvftoms.py]: ah good catch I forgot to remove that one

**Comment by bennahugo** [comment]
Alright the comments were incorporated apart from changing the quotes around field names, since these may contain spaces I find quotes clearer

**Comment by bennahugo** [comment]
@rubyvanrooyen this branch may come in handy as you are struggling with timeout and staging issues on 32k data.

@ludwigschwardt can we at some point get a version of this merged? - I see the branch has again gone out of date, however these features are a bit of a life saver if you are working with multi-target / multi-pointing or long observations which can be processed in batches.

**Comment by bennahugo** [comment]
@rubyvanrooyen there is an option to select scans which may help you in further chunking up the data for physical shipping

**Comment by ludwigschwardt** [comment]
Hi @bennahugo @rubyvanrooyen,

Sincere apologies for leaving this one hanging... It's time to give katdal some lovin' again! 😄 

I've done all flake8 fixes and reordered things a bit. Please let me know if it is still doing what you expect. The merge should be clean, since master has not touched `mvftoms.py`.

**Comment by ludwigschwardt** [comment]
Hi @bennahugo, if you are happy with the changes I made, I can merge this.

**Comment by bennahugo** [comment]
Thanks @ludwigschwardt I think this looks fine to me now. Sorry I missed this comment - it was on the day I went on leave and it got mangled in my inbox

**Comment by bennahugo** [comment]
Let me check one last thing though first - I noticed the scan selection switch didn't work when no scans were specified the other day. I got so used to selecting only some scans that I completely forgot about it.

**Comment by bennahugo** [comment]
Alright @ludwigschwardt. The issue stemmed from a simple python3 issue. I really wish map returned a cyclic reference - one of those gotchas that bites every now and again

**Comment by ludwigschwardt** [comment]
@bennahugo, have a look at the latest commits. I stripped out some `future` machinery that snuck back in (we are Python 3.6+ now), added f-strings and refactored the casa int list function a bit.

**Comment by bennahugo** [comment]
@ludwigschwardt Thanks looks good to me. I didn't want to use format strings because I didn't know which version of python you want to keep backwards compatible to - I know 16.04 only ships 3.5 so this sort of breaks easy backwards compatibility (and then there is the recent 2.7 backwards compatible merge - not sure for how long we can keep the two compatible. I know breaking changes in DASK has come into play since December and more are on the way with annotations?).

**Comment by ludwigschwardt** [comment]
We have the following setup:
- The `python2` branch and release 0.16.x - last support for Python 2
- Release 0.17.x - last support for Python 3.5
- `master` branch, release 0.18+ - Python 3.6+

We are targetting Python 3.6+ because I'm working towards the New Katpoint, which depends on Astropy 4.1 for satellite support and that needs 3.6.

Since you merged February's master into your branch, I assume you are also targetting 3.6+. That already pulled in masses of f-strings into `mvftoms.py` 😂 

If someone needs this functionality in Python 3.5, we could go back and cherry-pick some of the commits and make another branch on 0.17.x.

**Comment by bennahugo** [comment]
> We have the following setup:
> 
>     * The `python2` branch and release 0.16.x - last support for Python 2
> 
>     * Release 0.17.x - last support for Python 3.5
> 
>     * `master` branch, release 0.18+ - Python 3.6+
> 
> 
> We are targetting Python 3.6+ because I'm working towards the New Katpoint, which depends on Astropy 4.1 for satellite support and that needs 3.6.
> 
> Since you merged February's master into your branch, I assume you are also targetting 3.6+. That already pulled in masses of f-strings into `mvftoms.py` joy
> 
> If someone needs this functionality in Python 3.5, we could go back and cherry-pick some of the commits and make another branch on 0.17.x.

not at all if peeps need the features backported they can do that themselves - fully agree with you here. Onwards and upwards.

**Comment by ludwigschwardt** [comment]
Also, remember that Python 3.5 reached EOL in September 2020:

> Python 3.5 has now reached its end-of-life and has been retired. No more releases will be made.

**Review by ludwigschwardt — APPROVED** [review]
Good to go!

**Pull request merged**